### PR TITLE
Mqtt feature

### PR DESCRIPTION
**Pull request by pierrejochem**
Mqtt sending feature

### CODE REVIEW

File: src/Interfaces/MqttTransport.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?php
+namespace ErrorHeroModule\Interfaces;
```
Comment (samsonasik) [inline src/Interfaces/MqttTransport.php]: if https://github.com/McFizh/libMQTT is the lib you refer, then the service should be an instance of `LibMQTT\ Client`class in documentation config. No need to add this interface.

File: config/error-hero-module.local.php.dist [hunk 2]
```diff
@@ -106,5 +106,13 @@ json
                 'developer2@foo.com',
             ],
         ],
+        'mqtt-notification-settings' => [
+            // set to true to activate mqtt notification on log error event
+            'enable' => true,
```
[the remark this answers is not in the document]
Comment (pierrejochem) [reply config/error-hero-module.local.php.dist]: Sorry my fault. Will correct it.

File: config/expressive-error-hero-module.local.php.dist [hunk 3]
```diff
@@ -104,6 +104,14 @@ json
                 'developer2@foo.com',
             ],
         ],
+        'mqtt-notification-settings' => [
+            // set to true to activate mqtt notification on log error event
+            'enable' => true,
```
[the remark this answers is not in the document]
Comment (pierrejochem) [reply config/expressive-error-hero-module.local.php.dist]: Sorry my fault. Will correct it.

File: src/Interfaces/MqttTransport.php (hunk 1 above)
Comment (pierrejochem) [reply src/Interfaces/MqttTransport.php]: That is the one I use. My idea was to make it more prexible and not dependent to this lib. I could add it to the composer.json file as mandatory. Okay?

File: src/Handler/LoggingFactory.php [hunk 4]
```diff
@@ -49,6 +51,9 @@ public function __invoke($container)
         $mailMessageService   = null;
         $mailMessageTransport = null;
 
+        $mqttConfig = $errorHeroModuleLocalConfig['mqtt-notification-settings'];
```
[the remark this answers is not in the document]
Comment (pierrejochem) [reply src/Handler/LoggingFactory.php]: That is not 100% clear for me. Could you give me a hint please?

File: src/Handler/Logging.php [hunk 5]
```diff
@@ -104,6 +111,8 @@ public function __construct(
         $this->mailMessageTransport  = $mailMessageTransport;
         $this->emailReceivers        = $errorHeroModuleLocalConfig['email-notification-settings']['email-to-send'];
         $this->emailSender           = $errorHeroModuleLocalConfig['email-notification-settings']['email-from'];
+        $this->mqttMessageTransport = $mqttMessageTransport;
+        $this->mqttTopic = $errorHeroModuleLocalConfig['mqtt-notification-settings']['topic'];
```
Comment (samsonasik) [inline src/Handler/Logging.php]:
```php
$this->mqttTopic = isset($errorHeroModuleLocalConfig['mqtt-notification-settings']['topic']) ?
       $errorHeroModuleLocalConfig['mqtt-notification-settings']['topic'] :
       null;	
```

File: src/Handler/LoggingFactory.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (pierrejochem) [reply src/Handler/LoggingFactory.php]: Okay :-) Have not seen this for the email functions, but this make of course sense 👍

File: src/Interfaces/MqttTransport.php (hunk 1 above)
Comment (samsonasik) [reply src/Interfaces/MqttTransport.php]: you can register `mcfish/libmqtt` to composer.json under `require-dev` to allow you write test under `spec` directory, and add under `suggest` to suggest user to require when they want to use mqtt service

**Comment by samsonasik** [comment]
missing test and need to have updated/added at `spec` directory, I'm using [kahlan](https://github.com/kahlan/kahlan)

**Comment by pierrejochem** [comment]
Working on the changes and will add the tests

**Comment by pierrejochem** [comment]
Could you give me an advise for kahlan please? I've search for something for the email functionality so I can orientate on it or use it as template for the mqtt stuff, but can not find it.

**Comment by samsonasik** [comment]
you can create new integration test for it with new fixture config and new integration via error preview class, for example, on enable email:

- config fixture with `mail-message` and `mail-transport` definition that services defined under `service_manager`:  https://github.com/samsonasik/ErrorHeroModule/blob/8990c518bc4b2dcf356a45da3ed327ed670622b3/spec/Fixture/config/autoload-with-enable-sendmail/error-hero-module.local.php

- integration class: https://github.com/samsonasik/ErrorHeroModule/blob/8990c518bc4b2dcf356a45da3ed327ed670622b3/spec/Integration/IntegrationViaErrorPreviewControllerWithEnableSendMailSpec.php

To do this, you need to add `"mcfish/libmqtt": ^1.0.1` to `require-dev` for using Client class (testing purpose) and `suggest` to suggest when user want to use mqtt notification.

**Comment by samsonasik** [comment]
after more thinking about it, I think you can just create custom Writer, register service under `log_writers`, then add the writer under `log` -> `ErrorHeroModuleLogger` -> writers`, hence I close it ;)

**Comment by pierrejochem** [comment]
I will check this out. Thank you. I had my code finished, but problems with kahlan

**Pull request closed without merging**